### ISSUE TITLE
Review Metrics Support

### ISSUE THREAD

**Issue by pedroigor** [keycloak/keycloak#9036]
### Discussed in https://github.com/keycloak/keycloak/discussions/8490

<div type='discussions-op-text'>

<sup>Originally posted by **thomasdarimont** September 27, 2021</sup>
The book “Distributed Systems Observability'' by Cindy Sridharan describes logs, distributed tracing and metrics as 
essential telemetry types to monitor an application in production, which are also known as the “three pillars of observability”.
Currently Keycloak does not provide metrics out of the box and users who want to have metrics need to use extensions like 
the [aerogear keycloak-metrics-spi](https://github.com/aerogear/keycloak-metrics-spi) or implement their own metrics collection based on the smallrye-metrics support provided by the Wildfly and JBoss EAP runtimes. 
It would be very helpful for operations teams if Keycloak had a compelling set of useful metrics built-in.

The goal of this discussion is to shape the metrics part of Keycloak’s observability story with focus on Keycloak.X and to compile the foundation for a new metrics design document.

# Metrics

A metrics based monitoring of a Keycloak system could consist of interesting metrics that are relevant for different 
audiences like operations and SRE, as well as product teams.

Some of those metrics provide information about different layers of a system, including:

- [x] Process: CPU, Memory consumption, open file descriptors
- [x] JVM: memory, threads, classloading, metadata (java version)
- [x] Datasources: connection pool stats, metadata (database version)
- [x] HTTP server: request count per path/status code /latency distribution
- [x] JGroups: cluster communication stats
- [x] Infinispan: cache stats
- [ ] ~Integrations: outbound http request count / latency distribution~ -> consider tracing instead to analyze latencies and errors.
- [ ] Server: metrics collection duration, metadata (keycloak version) 
- [x] Keycloak: authentication stats, authorization stats
- [ ] Keycloak: inventory stats

# Keycloak Metrics

The application layer of a Keycloak system can provide many different metrics that could be arranged in a set of 
logical domains. Some of the following metrics might be coarse grained while others could be broken down further 
by additional context data, e.g. realm, error_code,  client_id, authenticator_execution, or protocol.

The following list serves as an example for high-level metrics that could theoretically be provided by Keycloak 
at some point in time.

The metrics listed below are based on [an earlier discussion about a compilation of metrics for Keycloak](https://docs.google.com/spreadsheets/d/1S-Xhsnw3BeYuAB5MtcbJDkjXBbJ73cuNrESxbvOyhKc/edit#gid=0).

## Model Metrics
Represents the system inventory, and denotes how many items of a particular type exist in the system.
This helps to keep an eye on the growth of the system.

Example metrics:
- #Realms
- #Users per Realm
- #Clients per Realm
- #Groups per Realm
- #Scopes per Realm

## Authentication Metrics
Represents authentication activity for users and clients.

Example metrics:
- #Logins
- #Login Errors
- #Logouts
- #Logout Errors
- #Login duration histogram
- #Client Login
- #Client Login Errors
- #Required Action Executions
- #Required Action Errors
- #Unique AuthenticationFlowSequence Executions (Username -> Password -> 2FA vs. Username -> Password)

## Authorization Metrics
Represents Authorization activity collected for the authz services.

Example metrics:
- #Access Requested
- #Access Granted
- #Access Denied

## User Metrics
Represents information about users and their metadata.

Example metrics:
- #Users by realm
- #Users by status blocked / locked / disabled
- #User with missing information (email, phoneNumber, address)
- #User with unverified information (email, phoneNumber, address)
- #Distribution of credentials
- #Groups by realm
- #Consents by client / type
- #New Users in interval (yesterday, last week, last month, last year)

## Client Metrics
Represents information about clients and their metadata.

Example metrics:
- #Clients by realm / protocol  / type / enabled / disabled

## OIDC Protocol Usage Metrics
Usage information about the OIDC protocol

Example metrics:
- #Token Requests
- #Token Request Errors
- #Refreshes
- #Refresh Errors
- #UserInfo Requests
- #UserInfo Request Errors
- #Token Exchanges
- #Token Exchanges Errors
- Token generation duration distribution by token type (by protocol mapper?)
- UserInfo generation duration distribution (by protocol mapper?)

## SAML Protocol Usage Metrics
Usage information about the SAML protocol

Example metrics:
- #AuthnRequests
- #AuthnRequest Errors
- Assertion generation duration distribution (by protocol mapper?)

## Federation Metrics
Information about user federation

Example metrics:
- #User lookups in storage
- #User lookup errors in storage

## Identity Brokering Metrics
Information about Identity Brokering

Example metrics:
- #Brokered user logins
- #Brokered user login errors

## Inbound / Endpoint Metrics
- #Inbound (HTTP) request/response by status / path / protocol
- #Inbound (HTTP) request/response latency distribution

In micrometer those are usually captured by the dimensional metric `http.server.requests{uri=...,status=...,...}`.

## Outbound Metrics
- #Outbound request/response  by status / path / protocol / destination
- #Outbound request/response latency distribution

In micrometer those are usually captured by the dimensional metric `http.client.requests{uri=...,status=...,...}`.

## Instance Metrics
Represents general information and metadata about the server. 
Some of those “metrics” are just simple gauges with a dummy value that exposes the actual metadata via labels.

Example metrics:
- Server Version
- Enabled features
- Metrics Collection duration
- #Exceptions by realm / exception class / cause

# Metrics Infrastructure
The Wildfly and JBoss EAP based Keycloak / RH SSO distributions use SmallRye metrics for their runtime metrics collection.
However the [Quarkus team recommends using micrometer](https://quarkus.io/blog/micrometer-metrics/) for a while now for custom metric collection. In order to follow this approach we will focus on micrometer based metrics for the new metrics support in Keycloak.X.

OS, Process and JVM based metrics are usually provided by the base metric libraries. 
In our case the micrometer library provides a set of useful JVM and system metrics out of the box: https://micrometer.io/docs/ref/jvm
The [micrometer Keycloak metrics SPI](https://github.com/micrometer-metrics/micrometer-keycloak/blob/master/src/main/java/io/micrometer/keycloak/MeterRegistryHolder.java#L38) provides some additional metrics that could be useful.

## Metrics instrumentation
Keycloak provides several ways to collect metrics synchronously, e.g.: event listeners, JAX-RS / container specific filters and HTTP client interceptors. Metrics that are more expensive to compute could be collected 
asynchronously by a dedicated metrics service that can execute datastore specific queries. 

Collected metrics could either be directly stored in the micrometer metric registry or buffered in an own data structure that periodically releases the metrics into an underlying registry.

Explicitly computed metrics could be represented as `Gauges` that are explicitly updated. 
Counted metrics like number of logins or failed logins could be recorded via `Counters` that are updated via 
event listeners or request filters / interceptors. 
Metrics around HTTP request processing should capture information about the request path, status code
and request durations. Additionally request duration recording should allow to track latency profiles.

Keycloak could provide components that enable metrics collection on multiple levels:
- MetricsEventListener: an event listener could update metrics based on user events or admin events 
- MetricsFilter: a server request / response filter could update request specific metrics
- MetricsInterceptor: a client request / response filter could update request specific metrics
- MetricsCollector: a metrics service could periodically run a compute metrics based on registered computation rules

Those metric components should access a shared metric registry, which holds the metadata and state that is eventually exposed by dedicated metric endpoints.

# Keycloak Metrics

## Initial Metrics Selection
Although many of the metrics mentioned above provide valuable insights about a Keycloak system, we should focus 
on a small initial subset of metrics that are provided out of the box.

## Built-in Metrics
Some core metrics should be built-in to Keycloak and provide some configuration options, like whether the metric is collected at all, or the granularity, e.g. additional tags, labels to add.

## Custom Metrics
Some of the metrics mentioned above could be provided out of the box by Keycloak, however there will be use-cases that can not be foreseen, which require the ability to contribute custom metrics to the system. 
For this Keycloak needs to provide a metrics SPI that enables users to add their own custom metrics.

## Metrics Configuration
We should have a way to let users control which metrics are collected / tracked by Keycloak.
Users should be able to control things like:
- Which metrics to enable?
- Which tags to emit alongside the metric?
- Which context information to include? (e.g. request parts or selected parameters)

## Exposing Metrics
Metrics need to be accessible for metric collection tools like Prometheus or InfluxDB. Those tools usually fetch metrics information from an HTTP endpoint. For this we could either provide one global metrics endpoint for the whole server and all realms or realm specific endpoints that can be consumed by the collectors. This model is supported by Quarkus out of the box via the `/q/metrics` global endpoint. This endpoint could then contain information about the process, jvm, instance, as well as all the Keycloak application metrics.

However in environments where a Keycloak system is shared among multiple different parties, e.g. a collection of realms per tenant model, users might only be allowed to access a subset of the metrics information via realm specific endpoints that provides only metrics for a particular realm. In this case an endpoint like `/auth/realm/$myrealm/metrics` could be used as a realm specific endpoint that only provides the Keycloak application metrics and perhaps a small subset of server metadata.

Note, that it should be possible to protect the endpoints which expose realm metrics.

## Metrics SPI

A metrics SPI should allow users to contribute new metrics to the Keycloak metrics collection.
The registered metrics could hook into the metrics collection infrastructure described above.

# Links
- [Aerogear Metrics SPI](https://github.com/aerogear/keycloak-metrics-spi)
- [Micrometer Keycloak Integeration](https://github.com/micrometer-metrics/micrometer-keycloak)
- [Custom Keycloak Metrics collection example](https://github.com/thomasdarimont/keycloak-project-example/tree/main/keycloak/extensions/src/main/java/com/github/thomasdarimont/keycloak/custom/metrics)

---

# Questions

- Which metrics are important for you? What other metrics would you like to see?
- What questions do you want to solve based on metrics and which metrics would support you here?

</div>

**Comment by ahus1** [comment]
@vmuzikar - I agree that Micrometer should be the way forward as it is recommended by the Quarkus team. 

I'd start with activity based metrics (number of logins, number of failed logins) as they would be in-memory counters and reset on Keycloak restart. They would then be aggregated and persisted via the metrics stack over all instances.  

I would skip state based metrics (like number of realms, or number of users per realm) as they would require reading data from the store which is expensive. 

If we have metrics per realm, this could lead to an explosion of metrics when there are a lot of realms (aka "metrics explosion"). Therefore I'd think I'd start without, or make this configurable.

Asking for timings, it would be great to switch between a accumulated time (just one counter per metric) to a histogram. While histograms have more information and allow for optimizations in the long tail, the accumulated time is a lot more efficient. 

In a nutshell, this is how I would start: 
* Migrate to Micrometer
* Start with in-memory counters for activities, broken down to OK and NOT OK results (if there are more than two status like 4-5, this is fine as well)
* Consider adding `realm` as a dimension to the metrics, and make it configurable.

**Comment by sschu** [comment]
@ahus1 Have you noticed https://github.com/keycloak/keycloak/pull/15235? And the best thing, it simply seems to work... ;)

**Comment by ahus1** [comment]
@sschu - yes, I have seen it, and I second the approach, although I haven't tested it.

Still I leave it for those people who're working on the Keycloak Quarkus to review and test it.

I see that it still carries the "DRAFT" in its subject, maybe it is now time to remove that?

**Comment by ahus1** [comment]
@sschu - and as usual, we like our PRs to have only a single commit. https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md#commit-messages-and-issue-linking

**Comment by sschu** [comment]
@ahus1 Of course (about the commit). I was first interested in some general feedback for this idea. There is going to be a meeting next week on this topic, then we will find out if we finalise this PR or if more needs to be considered/done.

**Comment by pedroigor** [comment]
@ahus1 Looks like a good plan.

@sschu I really missed your proposal. At first glance, it is ready to go. I'm wondering if that will help the setup on `keycloak-benchmark` too, @ahus1?

Shall we create a epic/umbrella issue to track the remaining work:

* Start with in-memory counters for activities, broken down to OK and NOT OK results (if there are more than two status like 4-5, this is fine as well)

I'm not sure about the realm dimension. IMO, we should focus on the server metrics. Perhaps look at realms in another iteration.

**Comment by sschu** [comment]
@pedroigor I agree, I would first go for the server metrics. My main question is if we should just go with what micrometer / infinispan deliver with the risk of metrics changing on updates or if the goal is to guarantee a stable metrics interface at the cost of additional effort. My favourite would be micrometer / infinispan delivering stable metrics though.

**Comment by pedroigor** [comment]
I see the benefits of defining our core metrics but I think those components know more than us what are their defaults and we should just leverage that.

Perhaps what we can do, at least initially, is to provide a test to make sure a snapshot of the metrics as they are now is there. So that we can quickly react in the future if something changed. If that is worth the cost ... I'm not sure if they are going to change so often.

**Comment by ahus1** [comment]
@pedroigor: Looking at the performance tests, I'm good with any kind of metrics. For now I'd vote for "the more the better" for built-in metrics, like for Infinispan. I wouldn't start to re-map them until we know which are essential to monitor a Keycloak instance. I'd assume there would be ready-made Infinispan Dashboards for Grafana that we can re-use. 

For performance tests I already have the response times and error rates on the Gatling side of things. Those server side metrics (count broken down to ok and error, eventually durations) would be necessary for a production system only, and we would the experiment with them in a load test scenario. 

And yes, let's defer the realm dimension for now.

@sschu - Let's consider all metrics experimental first until we decide they are ready.

**Comment by trixpan** [comment]
@ahus1 @pedroigor @sschu  are there any plans to work on this? The Aerogear metrics SPI is largely dormant with many of the PRs still waiting for @pb82 to review. Last issues to be closed dates back to mid 2023.

It seems unfair to place all the onus of metrics on Peter's back and the project should own the task of providing metrics.

**Comment by vmuzikar** [comment]
@trixpan I can confirm improved metrics support is still something we plan on doing. However, given its scope I can provide any specific details how it will look like, it will be probably more of a continuous effort based on smaller incremental steps. Please note there are already [basic metrics](https://www.keycloak.org/server/configuration-metrics) in Keycloak.

**Comment by ahus1** [comment]
There is now a PR https://github.com/keycloak/keycloak/pull/33046 which would allow to expose metrics based on user events, which would provide numbers for example on successful and unsuccessful logins.
Please review and comment over there.

**Comment by ahus1** [comment]
An update to those following this issue: 

In Keycloak 25, we added HTTP inbound metrics together with metrics for Infinispan and JGroups, that's why those items are checked-off in the TODO list at the top. 


The PR to add metrics for Keycloak events has now been merged to main, There is one more follow-up in #34602. It will be a preview-feature in KC26.1. A preview of the docs are available here: https://www.keycloak.org/nightly/server/event-metrics

Please provide your feedback to the parent epic: #34503. The more feedback we get that you tried it out and that it works for you as expected, the faster it can move from preview to supported.

**Comment by ahus1** [comment]
For the upcoming KC 26.1 release, OpenTelementry tracing is now fully supported and recorded all outgoing calls (SQL, LDAP, HTTP), see https://github.com/keycloak/keycloak/issues/34710

Assuming that Service Level Indicators for Keycloak can be derived as described in and any troubleshooting for latencies and errors can be done using tracing (see nightly docs https://www.keycloak.org/nightly/observability/keycloak-service-level-indicators), I see less of a need to additional metrics on outgoing calls, especially as those metrics lack a connection to the observed user behavior. 

So for now, I'll mark those outbound metrics as "not planned" until someone convinces me otherwise. Technically those metrics might also be problematic as outgoing URLs don't necessarily use placeholders, which would be a requirement to have the URL as a label in the metric.

Depending on the feedback I receive here, I might close issue as IMHO the most valuable parts have been delivered. Additional would require their own discussions, and should be handled in separate issues.

**Comment by schewara** [comment]
It's nice to see that this topic is still being worked on and is making progress.

I just want do add some small thoughts and observations from my end to this discussion.

### Prometheus/OpenMetrics vs OpenTelemetry metrics

As we have all our data scraping and shipping done through the OpenTelemetry Collector, there are still some issues lurking around in the background, which should be considered moving forward.

1. metric names and description(help metadata) are causing problems when being processed through otlp.
   we mainly observed these with some jvm metrics coming from micrometer directly, where the data was then dropped in the end
2. with OpenTelemetry tracing enabled, it would also make sense and would be great to be able to configure a `micrometer-registry-otlp` to support a single communication channel for all observability signals

But yes, these points probably qualify for individual discussions.

**Comment by ahus1** [comment]
@schewara - yes, those should be handled as individual issues. (2) and possibly (1) as well would most likely involve the Quarkus team.

**Comment by ahus1** [comment]
As of #37389, the user event metrics are now supported for the upcoming KC 26.2 release.

For anything that is missing, please open individual issues that would allow for a dedicated discussion on each area. Once there are multiple issues around that, we might then re-group them in an epic. Please feel free to ping me in any of those issues to get my input.
